### PR TITLE
Enable must-throw for unsupported intrinsics under optimizations

### PR DESCRIPTION
**Pull request by tannergooding**
This should put less pressure on the doing repetitive work and should allow merging of the various must throw blocks. The general issue was noticed in https://github.com/dotnet/runtime/pull/128513#issuecomment-4596763144, where we had some size regressions due to cold blocks for throwing intrinsics not being merged as they callee was unique for each; even though we don't really care about it being in the stack trace (its optimized code, inlining and loss of that stack trace entry was already possible).

### CODE REVIEW

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -1204,11 +1204,17 @@ static NamedIntrinsic binarySearchId(CORINFO_InstructionSet isa,
...
-//    comp       -- The compiler
-//    sig        -- The signature of the intrinsic
-//    className  -- The name of the class associated with the HWIntrinsic to lookup
-//    methodName -- The name of the method associated with the HWIntrinsic to lookup
-//    enclosingClassName -- The name of the enclosing class of X64 classes
+//    comp                -- The compiler
+//    sig                 -- The signature of the intrinsic
+//    className           -- The name of the class associated with the HWIntrinsic to lookup
+//    methodName          -- The name of the method associated with the HWIntrinsic to lookup
+//    enclosingClassName  -- The name of the enclosing class of X64 classes
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/hwintrinsic.cpp]: Fixed.

File: src/coreclr/jit/importercalls.cpp [hunk 2]
```diff
@@ -11551,7 +11570,7 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
         }
     }
 
-    if (result == NI_Illegal)
+    if ((result == NI_Illegal) || (result == NI_System_Runtime_Intrinsics_Intrinsic))
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/coreclr/jit/importercalls.cpp]: Fixed.

**Comment by tannergooding** [comment]
CC. @AndyAyersMS. Provides a small TP improvement (`-0.03%`) and reduction in bytes allocated (`-0.04%`). Provides a larger reduction in inline count (`-0.20%`). It then picks up a few diffs, mostly on the Linux boxes where we test hardware without AVX512 support.

Most of the scenarios are ultimately handled via DCE, so this namely just avoids the duplicative work and more importantly ensures the exception we produce in the JIT actually matches what managed was producing, with the nit that R2R remains incorrect for an edge case and is dependent on a version bump

**Comment by tannergooding** [comment]
Test failures are https://github.com/dotnet/runtime/issues/129037

**Comment by AndyAyersMS** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=1451481&view=ms.vss-build-web.run-extensions-tab) seem pretty sparse -- I thought I had looked at the prior diff set and there were more (though almost all in test collections).

**Comment by tannergooding** [comment]
I was surprised at the sparseness as well, perhaps its SPMI specific and related to the R2R/NAOT limitation (needing a version bump to ensure the correct exception is handled) here?

**Comment by tannergooding** [comment]
Ah, @AndyAyersMS it looks like its because of missed contexts. Previously we returned `nullptr` and so we never queried the info for a given helper call for `Inliner`, only in the actual intrinsic (the `Inlinee` that fails). Now we expand that directly and so the query happens and fails the context.

I'm getting direct diffs locally and will share here in a bit.

**Comment by tannergooding** [comment]
This is indeed where a lot of it went. For tests, as an example:
```
Found 4140 files with textual diffs.
Total bytes of delta: -6325 (-0.01 % of base)
```

Some of the SPMI diffs look to be because the Linux binaries used by SPMI don't have the Arm64 stuff trimmed, but it is trimmed when done locally, I presume this is some artifact of us doing some cross coverage.

In places I've looked, we appear to be merging the multiple throw blocks as expected, with most secondary throws already being DCE so we only end up with one. The two block case is mostly from places where we had two separate inlinees, each that has such a throw case.

**Review by AndyAyersMS — APPROVED** [review]
